lib/ts/ink_history.h

```diff
@@ -0,0 +1,82 @@
+/** @file
+
+  A brief file description
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#ifndef _ink_history_h_
+#define _ink_history_h_
+
+#include "ts/ink_platform.h"
+#include "ts/ink_defs.h"
+#include "ts/ink_apidefs.h"
+
+#define HISTORY_SIZE 64
+#define NO_REENTRANT -9999999
+struct HistoryEntry {
+  SourceLocation location;
+  unsigned short event = 0;
+  short reentrancy     = 0;
+};
+
+class History
+{
+public:
+  HistoryEntry history[HISTORY_SIZE];
+
+  void
+  add_history_entry(const SourceLocation &location, int event, int reentrant = NO_REENTRANT)
+  {
+    int pos                 = history_pos++ % HISTORY_SIZE;
+    history[pos].location   = location;
+    history[pos].event      = (unsigned short)event;
+    history[pos].reentrancy = (short)reentrant;
+  }
+
+  void
+  clear()
+  {
+    ink_zero(history);
+    history_pos = 0;
+  }
+
+  int
+  get_size()
+  {
+    return history_pos > HISTORY_SIZE ? HISTORY_SIZE : history_pos;
+  }
+
+  const char *
+  location_str(int index, char *buf, int length)
+  {
+    return history[index].location.str(buf, length);
+  }
+
+  const HistoryEntry &operator[](int i) { return history[i]; }
```

Comment:
All the functions that index into the history buffer should be unsigned, here and in the others. Also, the `history_pos` member should also be unsigned. There's no use case for signed indices here that I can see.